### PR TITLE
Add orchard binding_verification_key

### LINKED ISSUE #2102: Implement Orchard binding verification key derivation
Similar to `ShieldedData` for Sapling, we need to derive and validate the Orchard binding verification key.

### LINKED ISSUE #2103: Validate bindingSigOrchard for Transaction::V5
Very similar to how we've done this [in the `TransactionVerifier` service](https://github.com/ZcashFoundation/zebra/blob/74fd2a65fee23c89ddb2cea237cc8bc064c4fd03/zebra-consensus/src/transaction.rs#L305), for V5, we compute the binding verification key and use it to verify the binding_sig:

image
https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus

Related: 
https://github.com/ZcashFoundation/zebra/issues/1895
https://github.com/ZcashFoundation/zebra/issues/1939

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Binding verification key must be implemented.

Closes https://github.com/ZcashFoundation/zebra/issues/2102
Closes https://github.com/ZcashFoundation/zebra/issues/2103

### Specifications

https://zips.z.cash/protocol/protocol.pdf#orchardbalance

## Solution

Implement simular to sapling. This is a draft.

## Review

I will like @dconnolly to take an initial view. I made some comments where i have some doubts.

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2441)

### CODE REVIEW

File: zebra-chain/src/orchard/shielded_data.rs
Comment (oxarbitrage) [inline zebra-chain/src/orchard/shielded_data.rs]: The description is pretty much the same as sapling because i think it will work but we should probably not write the exact same thing.

File: zebra-chain/src/orchard/shielded_data.rs
Comment (oxarbitrage) [inline zebra-chain/src/orchard/shielded_data.rs:84]: We only have one `cv` field in orchard so i think this is correct but not sure.

File: zebra-consensus/src/transaction.rs
Comment (oxarbitrage) [inline zebra-consensus/src/transaction.rs]: We don't have test coverage here and also in a lot of validation functions in the same file such as `verify_sapling_shielded_data`, `verify_sprout_shielded_data`, `verify_v5_transaction`, `verify_v4_transaction`, etc.

How are we testing this validations ?

File: zebra-consensus/src/transaction.rs
Comment (oxarbitrage) [inline zebra-consensus/src/transaction.rs:552]: Assumes the async verifier will not fail but we maybe need to do the same as sapling. This is run themnot async and ignore the errors.

File: zebra-chain/src/orchard/shielded_data.rs
Comment (teor2345) [reply zebra-chain/src/orchard/shielded_data.rs]: This description needs to be updated based on the Orchard consensus rules:
https://zips.z.cash/protocol/protocol.pdf#actiondesc
https://zips.z.cash/protocol/protocol.pdf#orchardbalance
https://zips.z.cash/protocol/protocol.pdf#actionencodingandconsensus

Specifically:
* In some contexts, Orchard `Action`s replace Sapling `Spend`s and `Output`s
* In other contexts, Orchard input notes replace Sapling `Spend`s, and Orchard ouput notes replace Sapling `Output`s
* The key generation is different and needs to be based on the Orchard consensus rules

File: zebra-chain/src/orchard/shielded_data.rs
Comment (teor2345) [reply zebra-chain/src/orchard/shielded_data.rs:84]: Zebra's Sapling value balance code is probably wrong, see bug #1939.

We should implement this derivation based on the Orchard consensus rules:
https://zips.z.cash/protocol/protocol.pdf#orchardbalance

File: zebra-consensus/src/transaction.rs
Comment (teor2345) [reply zebra-consensus/src/transaction.rs]: We spoke about this at the Zebra meeting today:
* Transparent, Sprout, and Sapling code is validated using the mainnet and testnet chains, and some test vectors
* Orchard code needs test vectors from the ECC

There is also a specific follow up ticket for Sapling #2426.
Sprout tests are a low priority because it is deprecated.

File: zebra-consensus/src/transaction.rs
Comment (teor2345) [inline zebra-consensus/src/transaction.rs]: Please add TODO comments about the missing Orchard checks, with a ticket number for each check.

File: zebra-consensus/src/transaction.rs
Comment (teor2345) [reply zebra-consensus/src/transaction.rs:552]: If there are any validation errors, we should fix them, then merge this PR.

That probably means that this PR is blocked on the sapling value balance fix in #1939.

**Review by teor2345 — DISMISSED** [review]
It looks like we need to do two things before we can make progress on this PR:
1. Fix the sapling value balance errors in #1939 
2. Work with a cryptographer to understand the Orchard value balance consensus rules, and how they are different from Sapling

File: zebra-chain/src/orchard/shielded_data.rs
Comment (teor2345) [reply zebra-chain/src/orchard/shielded_data.rs]: This seems to be a relevant part of the spec:

> Consistency of vbalanceOrchard with the value commitments in Action descriptions is enforced by the Orchard binding signature. The rôle of this signature in the Orchard protocol is to prove that the net value spent (i.e. the total value spent minus the total value produced) by Action transfers is consistent with the vbalanceOrchard field of the transaction.
> Non-normative note: The other rôle of Sapling binding signatures, to prove that the signer knew the randomness used for commitments in order to prevent them from being replayed, is less important in Orchard because all Action descriptions have a spend authorization signature. Still, an Orchard binding signature does prove that the signer knew this commitment randomness; this provides defence in depth and reduces the differences of Orchard from Sapling, which may simplify security analysis.
> Instead of generating a key pair at random, we generate it as a function of the value commitments in the Action descriptions of the transaction, and the Orchard balancing value.

https://zips.z.cash/protocol/protocol.pdf#orchardbalance

**Comment by teor2345** [comment]
We need to merge PR #2463 before we merge this PR, so we have transaction verifier test coverage.

**Comment by teor2345** [comment]
We need to merge PR #2472 before merging this PR, so we're sure that the common Sapling/Orchard binding signature code works.

**Comment by dconnolly** [comment]
> We need to merge PR #2472 before merging this PR, so we're sure that the common Sapling/Orchard binding signature code works.

#2472 and #2463 have been merged so we can move forward with this PR i think

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [reply zebra-chain/src/orchard/shielded_data.rs:84]: We found the proximate source of the bug in our implementation of sighash, so I feel confident that this value balance code is actually correct now 👍

**Comment by teor2345** [comment]
> > We need to merge PR #2472 before merging this PR, so we're sure that the common Sapling/Orchard binding signature code works.
> 
> #2472 and #2463 have been merged so we can move forward with this PR i think

Also our cached state tests actually test this code now!

File: zebra-chain/src/orchard/shielded_data.rs
Comment (teor2345) [reply zebra-chain/src/orchard/shielded_data.rs:84]: Yep #1939 is fixed.

**Comment by dconnolly** [comment]
@oxarbitrage I moved this from draft to not-draft if that's ok with you

**Comment by teor2345** [comment]
> @oxarbitrage I moved this from draft to not-draft if that's ok with you

Can we focus on state changes for this sprint?

We can do this validation next sprint, or delay it until we have time.
But we need a bunch of the state changes sorted to move forward with the mempool.

**Comment by dconnolly** [comment]
> > @oxarbitrage I moved this from draft to not-draft if that's ok with you
> 
> Can we focus on state changes for this sprint?
> 
> We can do this validation next sprint, or delay it until we have time.
> But we need a bunch of the state changes sorted to move forward with the mempool.

Oh sure yes 👍

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/orchard/shielded_data.rs]:
```suggestion
    /// Getting the binding signature validating key from the Action description 
    /// value commitments and the balancing value implicitly checks that the 
    /// balancing value is consistent with the value transferred in the
    /// Action descriptions, but also proves that the signer knew the
    /// randomness used for the Action value commitments, which
    /// prevents replays of Action descriptions that perform an output.
    /// In Orchard, all Action descriptions have a spend authorization signature,
    /// therefore the proof of knowledge of the value commitment randomness
    /// is less important, but stills provides defense in depth, and reduces the 
    /// differences between Orchard and Sapling.
```

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/orchard/shielded_data.rs]:
```suggestion
    /// The net value of Orchard spends minus outputs in a transaction
    /// is called the balancing value, measured in zatoshi as a signed integer
    /// cv_balance.
```

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/orchard/shielded_data.rs]:
```suggestion
    /// Consistency of cv_balance with the value commitments in Action 
    /// descriptions is enforced by the binding signature.
```

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/orchard/shielded_data.rs]:
```suggestion
    /// Instead of generating a key pair at random, we generate it as a function
    /// of the value commitments in the Action descriptions of the transaction, and 
    /// the balancing value.
```

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [reply zebra-chain/src/orchard/shielded_data.rs]: @teor2345 I suggested several updates to these docs above 🎈

File: zebra-chain/src/orchard/shielded_data.rs
Comment (dconnolly) [inline zebra-chain/src/orchard/shielded_data.rs:85]: Looks good 👍

**Review by dconnolly — COMMENTED** [review]
Code looks good, some suggestions on the doc comments 📜

**Review by dconnolly — CHANGES_REQUESTED** [review]
Code looks good, some suggestions on the doc comments 📜

**Comment by mpguerra** [comment]
Does this no longer close #2103 also?

**Comment by oxarbitrage** [comment]
> Does this no longer close #2103 also?

Thank you for the question. I removed more code than i should in one of the rebases, the validation code was here so i will readd. With this addition this PR will close both tickets as initially intended.

**Review by oxarbitrage — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2441)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @dconnolly and @teor2345)

___
*[zebra-chain/src/orchard/shielded_data.rs, line 61 at r4](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI8ZnjrfltA9mmP-r4-61:-MhEZt6o78VtNCvzzsm0:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/42078878241fb3c0989d471b6e0387a627291a33/zebra-chain/src/orchard/shielded_data.rs#L61)):*
<details><summary><i>Previously, dconnolly (Deirdre Connolly) wrote…</i></summary><blockquote>

```suggestion
    /// Getting the binding signature validating key from the Action description 
    /// value commitments and the balancing value implicitly checks that the 
    /// balancing value is consistent with the value transferred in the
    /// Action descriptions, but also proves that the signer knew the
    /// randomness used for the Action value commitments, which
    /// prevents replays of Action descriptions that perform an output.
    /// In Orchard, all Action descriptions have a spend authorization signature,
    /// therefore the proof of knowledge of the value commitment randomness
    /// is less important, but stills provides defense in depth, and reduces the 
    /// differences between Orchard and Sapling.
```
</blockquote></details>

Done.
___
*[zebra-chain/src/orchard/shielded_data.rs, line 65 at r4](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI8ZnjrfltA9mmP-r4-65:-MhEZurgCHNUugLojUAE:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/42078878241fb3c0989d471b6e0387a627291a33/zebra-chain/src/orchard/shielded_data.rs#L65)):*
<details><summary><i>Previously, dconnolly (Deirdre Connolly) wrote…</i></summary><blockquote>

```suggestion
    /// The net value of Orchard spends minus outputs in a transaction
    /// is called the balancing value, measured in zatoshi as a signed integer
    /// cv_balance.
```
</blockquote></details>

Done.
___
*[zebra-chain/src/orchard/shielded_data.rs, line 69 at r4](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI8ZnjrfltA9mmP-r4-69:-MhEZvrcC4edvXq9__os:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/42078878241fb3c0989d471b6e0387a627291a33/zebra-chain/src/orchard/shielded_data.rs#L69)):*
<details><summary><i>Previously, dconnolly (Deirdre Connolly) wrote…</i></summary><blockquote>

```suggestion
    /// Consistency of cv_balance with the value commitments in Action 
    /// descriptions is enforced by the binding signature.
```
</blockquote></details>

Done.
___
*[zebra-chain/src/orchard/shielded_data.rs, line 73 at r4](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI8ZnjrfltA9mmP-r4-73:-MhEZwo20DhNswugjKa_:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/42078878241fb3c0989d471b6e0387a627291a33/zebra-chain/src/orchard/shielded_data.rs#L73)):*
<details><summary><i>Previously, dconnolly (Deirdre Connolly) wrote…</i></summary><blockquote>

```suggestion
    /// Instead of generating a key pair at random, we generate it as a function
    /// of the value commitments in the Action descriptions of the transaction, and 
    /// the balancing value.
```
</blockquote></details>

Done.
___
*[zebra-chain/src/orchard/shielded_data.rs, line 82 at r4](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI8ZnjrfltA9mmP-r4-82:-MhE_-dJ848uenj4tTtR:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/42078878241fb3c0989d471b6e0387a627291a33/zebra-chain/src/orchard/shielded_data.rs#L82)):*
<details><summary><i>Previously, dconnolly (Deirdre Connolly) wrote…</i></summary><blockquote>

Looks good 👍
</blockquote></details>

Done.
___
*[zebra-consensus/src/transaction.rs, line 544 at r1](https://reviewable.io/reviews/zcashfoundation/zebra/2441#-MfIOgI91V1AWGyBrbMg-r1-544:-MhE_41g8XtsQhxfvKRm:b-896fix) ([raw file](https://github.com/zcashfoundation/zebra/blob/a82b4fb6ccb706ea45cba22e069ce8ecae5a9993/zebra-consensus/src/transaction.rs#L544)):*
<details><summary><i>Previously, teor2345 (teor) wrote…</i></summary><blockquote>

Please add TODO comments about the missing Orchard checks, with a ticket number for each check.
</blockquote></details>

Done.

**Review by dconnolly — APPROVED** [review]
🎉

**Pull request merged**
